geps/gep-3440/index.md

```diff
@@ -0,0 +1,283 @@
+
+As with the `x-envoy-retry-grpc-on` header, the number of retries can be controlled via the `x-envoy-max-retries` header.
+
+By default, Envoy uses a fully jittered exponential backoff algorithm for retries.
+This means that after a failed attempt, Envoy waits a random amount of time (with jitter) based on
+an exponential growth pattern before trying again.
+- **Default Timing**: The base interval starts at 25ms, and each subsequent retry can increase
+this interval exponentially. By default, the maximum interval is capped at 250ms (10 times the base interval).
+- **Per-Attempt Timeout (`per_try_timeout`)**: Envoy allows you to set a specific timeout for each retry attempt,
+known as `per_try_timeout`. This timeout includes the initial request and each retry attempt.
+If you don’t specify a `per_try_timeout`, Envoy uses the global route timeout for the total duration of the request.
+
+In the Gateway API, this `per_try_timeout` will be equivalent to the BackendRequest timeout in the GRPCRouteRule.
+This ensures that each retry attempt, including the initial one, respects the overall timeout defined for the backend
+request, preventing retries from extending beyond the desired duration.
+
+### Nginx
+`ngx_http_grpc_module` in Nginx supports retries for gRPC requests using the `grpc_pass` directive.
+
+For gRPC requests, Nginx allows retries under certain conditions by forwarding requests to another server in
+an upstream pool when the initial request fails.
+The following configuration options are available to control when and how retries occur:
+1. **Retry Conditions** (`grpc_next_upstream`):
+    Nginx can retry a request if certain issues are encountered, such as:
+    - Network errors (e.g., connection or read errors).
+    - Timeouts when establishing a connection or reading a response.
+    - Invalid headers if the server sends an empty or malformed response.
+    - Specific HTTP error codes (e.g., 500, 502, 503, 504, 429) can be configured as retryable for gRPC responses.
+    By default, Nginx only retries on network error and timeout,
+    but you can specify other conditions (like HTTP status codes) to expand retry options.
+2. **Retry Limit by Time** (`grpc_next_upstream_timeout`):
+   You can set a total time limit for how long Nginx will attempt retries.
+   This limits the retry process to a specified time window, after which Nginx will stop attempting further retries.
+3. **Retry Limit by Number** (`grpc_next_upstream_tries`):
+    You can set a maximum number of retry attempts for a request.
+    Once this limit is reached, Nginx will stop attempting further retries.
+4. **Non-Idempotent Requests** (`non_idempotent`):
+    By default, Nginx does not retry non-idempotent requests (like POST or PUT) because they can cause side effects
+    if sent multiple times. However, you can enable retries for non-idempotent requests if needed.
```

Comment:
Does this imply that you MUST do something special to get NGINX to retry gRPC at all?

Reply:
By default, NGINX only retries gRPC requests on error and timeout.
To retry on other conditions (like http_503, invalid_header, or non_idempotent), you must explicitly set grpc_next_upstream with those options in your config.
https://nginx.org/en/docs/http/ngx_http_grpc_module.html#grpc_next_upstream